packages/timeline/src/browser/timeline-service.ts

```diff
@@ -0,0 +1,126 @@
+/********************************************************************************
+ * Copyright (C) 2020 RedHat and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable } from 'inversify';
+import { Disposable, Emitter, Event } from '@theia/core/lib/common';
+import { URI } from 'vscode-uri';
+import {
+    InternalTimelineOptions,
+    Timeline,
+    TimelineChangeEvent, TimelineItem, TimelineOptions,
+    TimelineProvider,
+    TimelineProvidersChangeEvent,
+    TimelineSource
+} from '../common/timeline-model';
+
+@injectable()
+export class TimelineService {
+    private readonly providers = new Map<string, TimelineProvider>();
+    private readonly providerSubscriptions = new Map<string, Disposable>();
+
+    private readonly onDidChangeProvidersEmitter = new Emitter<TimelineProvidersChangeEvent>();
+    readonly onDidChangeProviders: Event<TimelineProvidersChangeEvent> = this.onDidChangeProvidersEmitter.event;
+
+    private readonly onDidChangeTimelineEmitter = new Emitter<TimelineChangeEvent>();
+    readonly onDidChangeTimeline: Event<TimelineChangeEvent> = this.onDidChangeTimelineEmitter.event;
+
+    registerTimelineProvider(provider: TimelineProvider): Disposable {
+        const id = provider.id;
+
+        this.providers.set(id, provider);
+        if (provider.onDidChange) {
+            this.providerSubscriptions.set(id, provider.onDidChange(e => this.onDidChangeTimelineEmitter.fire(e)));
+        }
+        this.onDidChangeProvidersEmitter.fire({ added: [id] });
+
+        return Disposable.create(() => this.unregisterTimelineProvider(id));
+    }
+
+    unregisterTimelineProvider(id: string): void {
+        if (!this.providers.has(id)) {
+            return;
+        }
+
+        this.providers.delete(id);
+        this.providerSubscriptions.delete(id);
```

Comment:
Surely you should be disposing the listener here?  You make the effort to keep the listener disposables in `this.providerSubscriptions` but you don't do anything with them.

Reply:
I've fixed that by disposing the provider directly instead of storing disposables in a map.